### PR TITLE
Cli testing

### PR DESCRIPTION
**Pull request by elsapeng**

### CODE REVIEW

File: bin/swagger-project.js [hunk 1]
```diff
@@ -63,6 +65,14 @@ app
   .option('-b, --debug-brk [port]', 'start in remote debug mode, wait for debugger connect')
   .option('-m, --mock', 'run in mock mode')
   .action(execute(project.test));
+  
+app 
+  .command('generate-test [directory]')
+  .description('Generate the test template')
+  .option('-p, --path-name [path]', 'a sepecific path of the api, also suppport regular expression')
+  .option('-f, --test-module <module>', 'one of: ' + testmodules, 'supertest')
```
Comment (mohsen1) [inline bin/swagger-project.js]: Don't hardcode `"supertest"` and other defaults 
`.option('-f, --test-module <module>', 'one of: ' + testmodules, 'supertest')` => `.option('-f, --test-module <module>', 'one of: ' + testmodules,  project.testmodules[0])`

File: lib/commands/project/project.js [hunk 2]
```diff
@@ -33,6 +35,9 @@ var FRAMEWORKS = {
   sails:   { source: 'sails' }
 };
 
+var ASSERTIONTYPES = ['should', 'expect', 'assert'];
+var TESTMODULES = ['request', 'supertest'];
```
Comment (mohsen1) [inline lib/commands/project/project.js]: The default should go first

File: lib/commands/project/project.js [hunk 3]
```diff
@@ -363,3 +373,57 @@ function spawn(command, options, cwd, cb) {
...
 }
+
+
+//.option('-p, --path-name [path]', 'a sepecific path of the api')
+//.option('-f, --test-module <module>', 'one of: ' + testmodules)
+//.option('-t, --assertion-format <type>', 'one of: ' + assertiontypes)
+function testGenerate(directory, options, cb) {
+  var pathList = [];
+  var desiredPaths = [];
+  var swaggerSpec = require('../../util/spec');
```
Comment (mohsen1) [inline lib/commands/project/project.js]: should go to top, also in `verify` remove the extra `require`

File: lib/commands/project/project.js [hunk 4]
```diff
@@ -363,3 +373,57 @@ function spawn(command, options, cwd, cb) {
...
+
+  //read the yaml file and validate it
+  readProject(directory, options, function(err, project) {
+    if (err) { return cb(err); }
+    swaggerSpec.validateSwagger(project.api.swagger, options, cb);
```
Comment (mohsen1) [inline lib/commands/project/project.js]: put all your code from here in the callback function and call `cb` when everything is done.

File: lib/commands/project/project.js [hunk 5]
```diff
@@ -363,3 +373,57 @@ function spawn(command, options, cwd, cb) {
...
+      return cb(new Error(util.format('Unknown type: %j. Valid types: %s', options.testModule, TESTMODULES.join(', '))));
+    }
+
+    // check if the assertion-format is one of the three
+    if (options.assertionFormat && !_.includes(ASSERTIONTYPES, options.assertionFormat)) {
+      return cb(new Error(util.format('Unknown type: %j. Valid types: %s', options.assertionFormat, ASSERTIONTYPES.join(', '))));
+    }
+
+    // process the paths option
+    if (options.pathName){
```
Comment (mohsen1) [inline lib/commands/project/project.js]: Test this carefully.

File: lib/commands/project/project.js [hunk 6]
```diff
@@ -363,3 +373,57 @@ function spawn(command, options, cwd, cb) {
...
+      testModule: options.testModule,
+      assertionFormat: options.assertionFormat
+    };
+    var result = template.testGen(project.api.swagger, config);
+
+    //output the result
+    async.series(result, function(item, cb) {
+      var path = item.name;
+      var filename = path.substirng(1) + '.js';
+      fs.writeFile('./test/' + filename, item.test, function (err) {
```
Comment (mohsen1) [inline lib/commands/project/project.js]: Use the path module for putting together strings to make a path

@theganyo can you help us with path structure?

File: lib/commands/project/project.js [hunk 7]
```diff
@@ -363,3 +373,57 @@ function spawn(command, options, cwd, cb) {
...
+        if (err) return cb(err);
```
Comment (mohsen1) [inline lib/commands/project/project.js]: Simply pass the callback to `fs.writeFile`:

File: lib/commands/project/project.js [hunk 8]
```diff
@@ -363,3 +373,57 @@ function spawn(command, options, cwd, cb) {
...
+      });
+    });
+  });
+}
```
Comment (mohsen1) [inline lib/commands/project/project.js:437]: `;`?

File: package.json [hunk 9]
```diff
@@ -29,7 +29,9 @@
     "nodemon": "^1.3.7",
     "serve-static": "^1.9.2",
     "swagger-editor": "^2.9.2",
-    "swagger-tools": "^0.8.6"
+    "swagger-tools": "^0.8.6",
+    "async": "^1.2.1",
```
Comment (mohsen1) [inline package.json:33]: @theganyo should we use `async` or do you use something else for doing async map/each?

File: lib/commands/project/project.js (hunk 7 above)
Comment (elsapeng) [inline lib/commands/project/project.js]: changed.

File: lib/commands/project/project.js (hunk 4 above)
Comment (elsapeng) [inline lib/commands/project/project.js]: changed.

File: lib/commands/project/project.js (hunk 3 above)
Comment (elsapeng) [inline lib/commands/project/project.js]: changed.

**Pull request closed without merging**